### ISSUE TITLE
Chinese autokey skipped all chinese charaters 

### ISSUE THREAD

**Issue by arthurtalkgoal** [keystonejs/keystone#1366]
I set up the ***autokey*** of the model as follows:
```
var Deal = new keystone.List('Deal', {
    map: { name: 'name' },
	autokey: { path: 'slug', from: 'name', unique: true },
  track: true,
  defaultSort: '-createdAt'
});
```
When user input chinese name , the slug will skip all the chinese:
```
維他命E補濕系列、異國麝香、異國麝香沈木、白麝香、白麝香LIBERTINE
```
becomes
***e-e-libertine-***

**Comment by JedWatson** [comment]
This isn't great. Can you suggest a way to translate Chinese characters into URL-safe characters?

It looks like the [pinyin](https://www.npmjs.com/package/pinyin) package might do the trick...

**Comment by arthurtalkgoal** [comment]
Hi @JedWatson, 

worse is that some whole Chinese title will not be able to create the slug at all (it happens in most of my posts).

I know Rails have a very good slug gems doing it greatly:
https://github.com/norman/friendly_id

The related issue is: 
https://github.com/norman/friendly_id/issues/304

Some people want the pinyin, but it is better to do the friendly_id generic way - handle any non-Latin chars (instead of particular Chinese)

And they mention this function in the issue:
http://api.rubyonrails.org/classes/ActiveSupport/Inflector.html#method-i-transliterate

**Comment by JohnnyEstilles** [comment]
@JedWatson, I've been using a little module called [limax](https://github.com/lovell/limax) for this purpose. It currently support Chinese, Japanese, Cyrilic and Latin scripts. Use cases I've ran into that it doesn't support are Thai and Arabic (there's currently an issue open requesting Arabic). 

This is the most complete transliteration utility (at least for my use-cases) that I have found for node so far. Though I should say my search has been, by no means, exhaustive.

**Comment by arthurtalkgoal** [comment]
Do we all agree with @JohnnyEstilles suggestion.  limax.

If so, I can join the PR.

**Comment by JedWatson** [comment]
I do - having checked it out, limax looks great (although it has a really unfortunately autocorrect, btw - I just had to try 3 times to type that!)

PR would be very welcome implementing it. Thanks!

**Comment by arthurtalkgoal** [comment]
Hi, I updated the ***keystone-utils*** and ***keystone***.

Notice that I added the optional field "defaultLocale" for autokey.

Since I found that ***limax*** is bad on guessing Traditional Chinese (it usually failed or get Japanese).

**Comment by JedWatson** [comment]
@arthurtalkgoal before I merge them, are you interested in trying out https://www.npmjs.com/package/pinyin and seeing if it performs better?

(I would do this myself, but would be a terrible judge. My mandarin is about the level of a 3 year old :smile:)

**Comment by arthurtalkgoal** [comment]
Hi @JedWatson ,

Fact, limax has mixed the ***pinyin*** with other translation.  The ***hardest thing is guessing the language of the string***.  

limax use https://github.com/dachev/node-cld to do it. cld is from Google https://code.google.com/p/cld2/.  I tried the other detection, but cannot find any one works with  chinese.  cld2 works pretty good if ***languageHint*** is passed.

I think the other way for detection is from the req.locale.  However, it doesn't work before keystone is i18n (keystone header or cookies is not assumed to have the locale information, or the information can be used in models)

**Comment by JedWatson** [comment]
@arthurtalkgoal aah, I get it. Thanks for the clarification (also thanks for the PRs!)

I'm currently trying to figure out if there's a way we could make this an optional plugin for Keystone (I'm currently trying to break it down into smaller packages and plugins) but if not, will merge your PRs as they are so this can be resolved.

**Comment by arthurtalkgoal** [comment]
Hi @JedWatson ,

I only changed this function in keystone 
```
var slug = exports.slug = function(str, sep, options) {
}
```

Re trailing spaces, I am happy to do it. However, my atom editor automatically remove that. Let me do it when I back home.

And I found some other repository like spree/spree are very strict on that, they use the following to enforce no trailing spaces or lines:
https://github.com/houndci

**Comment by JedWatson** [comment]
@arthurtalkgoal I was thinking more adding the package to `keystone-utils`, but it's fine - let's go with it for now. I think the improvements (including unicode) are general enough to warrant the additional dependencies.

We've gradually been tightening the controls on coding practice, houndci is a good thing to look at including!

Thanks for your work on this.

**Comment by lovell** [comment]
Hello, `limax` maintainer (and `cld` contributor) here.

@arthurtalkgoal Are you able to open an issue against `limax` with some Traditional Chinese test cases that are currently failing detection? I suspect the underlying code in `cld` extracted from Chrome needs updating to pick up recent language-detection improvements.

**Comment by JedWatson** [comment]
Hi @lovell - thanks for jumping in and picking this up!

**Comment by lovell** [comment]
@JedWatson no worries, see https://github.com/dachev/node-cld/pull/14

**Comment by arthurtalkgoal** [comment]
Hi @lovell and @JedWatson ,

Does it mean we can go ahead this version and update when they move to github?

**Comment by arthurtalkgoal** [comment]
Hi @lovell,

I found a little problem with the limax pinyin, the translation include the ***tones numbers*** e.g.

```
弄堂里的菜品赤醬

nong4-tang2-li3-di2-cai4-pin3-chi4-jiang4
```

It is no good for slug purpose.  Not sure if limax need that number? 

If limax keep that number, let me update the keystone codes to remove the numbers.

**Comment by lovell** [comment]
Hi @arthurtalkgoal, are you suggesting `nong-tang-li-di-cai-pin-chi-jiang` would be the preferred transliteration here?

`limax` currently uses tone numbers because tone diacritics do not conform to RFC3986. When I first created the module a couple of years ago the use of tone numbers in URLs was an acceptable practice, based on feedback from native speakers. It's possible this is no longer the case, or at least might need to be configurable.

If so, let's move this discussion to [limax](https://github.com/lovell/limax/issues) and fix it there so Keystone and other dependents can take advantage.

**Comment by JedWatson** [comment]
Implemented by #1415

**Comment by andyhu** [comment]
You can also try [transliteration](http://github.com/andyhu/transliteration) module.
